common: refactoring transfer model.

oopsmonk opened the pull request:
This PR fixes #578.
* refactoring and adding helper functions and error handling in the transfer model.
* converting DOS to UNIX format in `common/model/transfer.c` file.

# Test Plan:
CI must pass.

Commit:
common: refactoring transfer model.

adding helper functions and error handling.

thibault-martinez commented on common/model/transfer.c:
```diff
@@ -1,676 +1,474 @@
+           sizeof(transaction->bundle));
+    memcpy(transaction->address, transfer->address,
+           sizeof(transaction->address));
+    memcpy(transaction->obsolete_tag, transfer->tag,
```
Shouldn't it be `transfer->obsolete_tag` ?

tsvisabo commented on common/model/transfer.c:
```diff
@@ -1,676 +1,474 @@
+                  __func__, __LINE__);
+    }
+  } else {
+    log_error(TRANSFER_LOGGER_ID, "[%s:%d] transfer type not match.\n",
```
do not

tsvisabo commented on common/model/transfer.c:
```diff
@@ -1,676 +1,474 @@
+                  __func__, __LINE__);
+    }
+  } else {
+    log_error(TRANSFER_LOGGER_ID, "[%s:%d] transfer type not match.\n",
```
do not

tsvisabo commented on common/model/transfer.c:
```diff
@@ -1,676 +1,474 @@
+  transfer_data_t* tf_data = NULL;
+  transfer_t* transfer = NULL;
+  if (!address) {
+    log_error(TRANSFER_LOGGER_ID, "[%s:%d] address cannot be NULL.\n", __func__,
```
i think "must not" is preferable

oopsmonk replied:
it's no difference in this case.

tsvisabo commented on common/model/transfer.c:
```diff
@@ -1,676 +1,474 @@
+  transfer_value_out_t* value_out = NULL;
+  transfer_t* tf = NULL;
+  if (!validate_output(output)) {
+    log_error(TRANSFER_LOGGER_ID, "[%s:%d] output is invalid.\n", __func__,
```
cannot -> must not

tsvisabo commented on common/model/transfer.c:
```diff
@@ -1,676 +1,474 @@
+
+void transfer_iterator_free(transfer_iterator_t** iter) {
+  transfer_iterator_t* iter_p = *iter;
+  if (*iter || iter) {
```
Did you mean `&&`, also, you need to verify `iter` is not null before dereferencing

tsvisabo commented on common/model/transfer.c:
```diff
@@ -1,676 +1,474 @@
+      transfer_iterator->current_transfer_transaction_index = 0;
+      return transfer_iterator_next(transfer_iterator);
+    }
+    transaction = transfer_iterator->transaction;
```
could you export that logic to a separate function?

oopsmonk replied:
I think this logic is here only.

tsvisabo replied:
yeah, it is not crucial, but it makes the code clearer

oopsmonk replied:
Why do you think it's not clear? I mean no complex logic and it has some comments here.

tsvisabo commented on common/model/transfer.c:
```diff
@@ -1,676 +1,474 @@
+      transfer_iterator->transfers_count) {
+    // Dynamically allocate a transaction structure if none was
+    // previously provided with transfer_iterator_set_transaction
+    if (!transfer_iterator->transaction) {
```
doesn't this method suppose to be a getter/read only?
should we not return null in that case?

oopsmonk replied:
what do you mean? could you give more details?

tsvisabo replied:
you are allocating a new transaction, should we do that when someone calls `next`?

oopsmonk replied:
it depends on the user, the comment on line 404 has answered your question.

thibault-martinez commented on common/model/transfer.c:
```diff
@@ -1,676 +1,474 @@
+
+#include "common/model/transfer.h"
+
+#define TRANSFER_LOGGER_ID "cclient_core_api"
```
Wrong logger id

thibault-martinez commented on common/model/transfer.c:
```diff
@@ -1,676 +1,474 @@
+
+  if (transfer->type == VALUE_IN) {
+    value_in = (transfer_value_in_t*)transfer->meta;
+    memcpy(tx->signature_or_message, value_in->data, sizeof(value_in->len));
```
`value_in->len` instead of `sizeof(value_in->len)` ?

thibault-martinez commented on common/model/transfer.c:
```diff
@@ -1,676 +1,474 @@
+}
+
+// create data transfer object.
+transfer_t* transfer_data_new(flex_trit_t const* const address,
```
What about `_init` functions that takes and fills a struct instead of `_new` functions that allocs and fills a struct ?

oopsmonk replied:
it allocates transfer_t.

thibault-martinez replied:
Yes, but why ? Is this really necessary ?
By taking a `transfer_t *` as parameter we allow the user to fill a static or a dynamic `transfer_t`. IMHO we should avoid dynamic allocation as much as possible in code used in the cclient.

oopsmonk replied:
this change just provides helper methods and keeps original design in this model. 
In my understanding, structures are mostly pointers in the transfer.

thibault-martinez replied:
Yes but since it will only be used by the client maybe it's a good thing to adapt it to the client needs. Pointers can be addresses of static structures. We'll probably need to remove dynamic allocation someday.

thibault-martinez commented on common/model/transfer.c:
```diff
@@ -1,676 +1,474 @@
+}
+
+void transfer_free(transfer_t** transfer) {
+  if (transfer || *transfer) {
```
`&&`

thibault-martinez commented on common/model/transfer.c:
```diff
@@ -1,676 +1,474 @@
+  return 0;
+}
+
+transfer_ctx_t* transfer_ctx_new() {
```
Do we need dynamic allocation for the ctx ?

oopsmonk replied:
not sure, but just keep the logic in the previous change.

thibault-martinez replied:
I think we don't need it

oopsmonk replied:
btw, why it exists?

thibault-martinez replied:
I have no idea :)

oopsmonk replied:
then, why do you think we don't need it?

thibault-martinez replied:
Looking where it's being used (e.g. `transfer_iterator_new`), I don't understand why it's being dynamically allocated

oopsmonk replied:
I prefer not to change something doesn't fully understand, we can seek for the answer from @th0br0 and @ifullgaz.

th0br0 replied:
Feel free to remove the method, it was an experiment. But in hindsight, I'm not that happy with the pointer aliases either.

thibault-martinez commented on common/model/transfer.c:
```diff
@@ -1,676 +1,474 @@
+bool transfer_ctx_init(transfer_ctx_t* transfer_ctx, transfer_t* transfers[],
+                       size_t len) {
+  size_t i;
+  int64_t total = 0;
```
Maybe set `transfer_ctx->count = 0` before the loop in case we re-init it ?

oopsmonk replied:
should not re-init an init method.

thibault-martinez replied:
Maybe put a comment then ?

thibault-martinez commented on common/model/transfer.c:
```diff
@@ -1,676 +1,474 @@
+    free((*transfer)->meta);
+    free(*transfer);
+    *transfer = NULL;
+  }
```
We should probably wipe out the `seed` field if the transfer is a `VALUE_OUT`

oopsmonk replied:
that is a pointer in `transfer_value_out_t`

thibault-martinez replied:
Fair enough

thibault-martinez commented on common/model/transfer.c:
```diff
@@ -272,7 +272,7 @@ transfer_t* transfer_value_in_new(flex_trit_t const* const address,
 
 void transfer_free(transfer_t** transfer) {
-  if (transfer || *transfer) {
+  if (*transfer) {
```
We still need to check `transfer` before dereferencing it

thibault-martinez commented on common/model/transfer.c:
```diff
@@ -383,7 +383,7 @@ transfer_iterator_t* transfer_iterator_new(transfer_t* transfers[], size_t len,
 void transfer_iterator_free(transfer_iterator_t** iter) {
   transfer_iterator_t* iter_p = *iter;
-  if (*iter || iter) {
+  if (*iter) {
```
We still need to check `iter` before dereferencing it

Commit:
common: fixing typo and grammatical errors.

Commit:
common: added a prarmeter to transfer_iterator_new.

checking transaction object when creating transfer instead of in the
iterations.

Commit:
Merge branch 'develop' into common_transfer
